File: 463_Island_Perimeter.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/island-perimeter/
# You are given a map in form of a two-dimensional integer grid where 1 represents land and 0 represents water.
# Grid cells are connected horizontally/vertically (not diagonally). The grid is completely surrounded by water,
# and there is exactly one island (i.e., one or more connected land cells). The island doesn't have "lakes" (water
# inside that isn't connected to the water around the island). One cell is a square with side length 1. The grid
# is rectangular, width and height don't exceed 100. Determine the perimeter of the island.

# Iterate over grid. If a cell is land, add 4 to perimeter assuming cell is surrounded by water. Then for each of lower
# and right cells that are land, decrement the perimeter by 2 (both sides of the internal edge).
# Time - O(mn)
# Space - O(mn)

class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        rows, cols = len(grid), len(grid[0])
        for row in grid:                # pad with water
            row.append(0)
        grid.append([0] * (cols + 1))
        perimiter = 0

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    perimiter += 4
                    if grid[r + 1][c] == 1:
                        perimiter -= 2
                    if grid[r][c + 1] == 1:
                        perimiter -= 2

        return perimiter

File: java/463_Island_Perimeter.java
/*
https://leetcode.com/problems/island-perimeter/
You are given a map in form of a two-dimensional integer grid where 1 represents land and 0 represents water. Grid
cells are connected horizontally/vertically (not diagonally). The grid is completely surrounded by water, and there
is exactly one island (i.e., one or more connected land cells). The island doesn't have "lakes" (water inside that
isn't connected to the water around the island). One cell is a square with side length 1.
Determine the perimeter of the island.

Iterate over grid. Land squares have 4 borders with water by default. If a cell has land to the left, reduce the
number of perimeters by 2. Likewise for land above.
Time - O(mn)
Space - O(1)
*/

public class Solution {
    public int islandPerimeter(int[][] grid) {

        int perimeter = 0;

        for (int row = 0; row < grid.length; ++row) {
            for (int col = 0; col < grid[0].length; ++col) {

                if (grid[row][col] == 1) {      // ignore if cell is blank
                    perimeter += 4;
                    if (row != 0 && grid[row - 1][col] == 1)
                        perimeter -= 2;
                    if (col != 0 && grid[row][col - 1] == 1)
                        perimeter -= 2;
                }
            }
        }

        return perimeter;
    }
}
